src/batou_ext/nixos.py

```diff
@@ -1,86 +1,75 @@
+import json
+import os
 from pathlib import Path
+from typing import Optional
 
 from batou.component import Component
-from batou.lib.file import Purge
+from batou.lib.file import JSONContent, ManagedContentBase, Purge
 
-from batou_ext.nix import NixFile, component_to_nix, nix_dict_to_nix
+import batou_ext.nix
 
-# XXX: error messages stemming from the "batouModule" are displayed with
-# wrong location information, it shows the glue module instead of the actual
-# module. Can we fix that somehow?
-GLUE_MODULE_TEMPLATE = """\
-{{ pkgs, ... }}@args:
 
-with builtins;
+class NixosModuleContent(ManagedContentBase):
+    json_file: Optional[Component] = None
 
-let
-  moduleArgs = (import {workdir}/{prefix}_generated_context.nix) // args;
-  batouModule = import {workdir}/{name}.nix moduleArgs;
-in
-{{ imports = [ batouModule ]; }}
-"""
+    def render(self):
+        self.content = self.expand(
+            """\
+# this file is managed by batou, any manual changes will be overwritten by the next deployment
+{pkgs, ...}@args: let
+  component = builtins.fromJSON (builtins.readFile {{component.json_file.path}});
+  module = (
 
+{{component.content}}
 
-class NixOSModuleContext(Component):
-
-    source_component: Component = None
-    prefix: str = None
-
-    def configure(self):
-
-        if self.source_component:
-            component = self.source_component
-        else:
-            component = self.parent
-
-        context = nix_dict_to_nix({"component": component_to_nix(component)})
-
-        if self.prefix is None:
-            self.prefix = component.__class__.__name__.lower()
-
-        self += NixFile(
-            f"{self.prefix}_generated_context.nix",
-            content=context,
-            format_nix_code=True,
-            sensitive_data=True,
+  );
+in {
+  imports = [(module (args // { inherit component; }))];
```

Comment:
fwiw, I think what I'd prefer here is actually

```nix
{ component, ... }:
{ config, lib, pkgs, ... }:
{

}
```

i.e. a function that generates a module given the component context instead of mixing this together with `_module.args` (if this ever gets a `component`-key, it will be funny).

I'm not sure if there's a transition path that's not just a pile of hacks, so we probably have no choice and need to leave this.

Reply:
Honestly, that is a better version. 

Transitioning would have to be manual but could just be part of any regular maintenance. The change is not too drastic that we'd have to account for it in this component.

tiny change but I'd change it to just 

```
component:
{ config, lib, pkgs, ... }:
{

}
```

since there are no other arguments to pass here

Reply:
> since there are no other arguments to pass here

fine by me.
This will require an entry in the release notes then, though.

Reply:
100%, this is a breaking change